### PR TITLE
fix(runtime): fix bash interrupt on program that cannot be stopped via ctrl+c

### LINKED ISSUE #4158: [Bug]: Runtime CmdRun timeout issue. Command execution failed: {"detail":"invalid literal for int() with base 10: \'echo\'"}
### Is there an existing issue for the same bug?

- [X] I have checked the troubleshooting document at https://docs.all-hands.dev/modules/usage/troubleshooting
- [X] I have checked the existing issues.

### Describe the bug

Sometime cmd execution will get error like this:
```
02:51:33 - ACTION
**CmdRunAction (source=EventSource.AGENT)**
THOUGHT: I will run the development server again in the background to test the autoreloader:
COMMAND:
python ./tests/i18n/sampleproject/manage.py runserver > server.log 2>&1 &
02:51:33 - openhands:INFO: runtime.py:411 - Awaiting session
02:51:33 - openhands:INFO: runtime.py:304 -
-----------------------------------Container logs:-----------------------------------
    |INFO:     172.17.0.1:33556 - "GET /alive HTTP/1.1" 200 OK
    |02:51:32 - openhands:ERROR: client.py:343 - Failed to execute command. Error: Timeout exceeded.
    |<pexpect.pty_spawn.spawn object at 0x7fc4c833d610>                                                                                          |command: /usr/bin/su
    |args: [b'/usr/bin/su', b'root']
    |buffer (last 100 chars): ".settings'\r\nStarting development server at http://127.0.0.1:8000/\r\nQuit the server with CONTROL-C.\r\n"
    |before (last 100 chars): ".settings'\r\nStarting development server at http://127.0.0.1:8000/\r\nQuit the server with CONTROL-C.\r\n"
    |after: <class 'pexpect.exceptions.TIMEOUT'>
    |match: None
    |match_index: None
    |exitstatus: None
    |flag_eof: False
    |pid: 55
    |child_fd: 8
    |closed: False
    |timeout: 30
    |delimiter: <class 'pexpect.exceptions.EOF'>
    |logfile: None
    |logfile_read: None
    |logfile_send: None
    |maxread: 2000
    |ignorecase: False
    |searchwindowsize: None
    |delaybeforesend: 0.05
    |delayafterclose: 0.1
    |delayafterterminate: 0.1
    |searcher: searcher_re:
    |    0: re.compile('\\[PEXPECT_BEGIN\\]\\s*(.*?)\\s*([a-z0-9_-]*)@([a-zA-Z0-9.-]*):(.+)\\s*\\[PEXPECT_END\\]')
--------------------------------------------------------------------------------
02:51:33 - openhands:ERROR: runtime.py:430 - Error from server: {"detail":"invalid literal for int() with base 10: 'echo'"}
02:51:33 - OBSERVATION
ErrorObservation(content='Command execution failed: {"detail":"invalid literal for int() with base 10: \'echo\'"}', observation='error')
```

We should write a test case try to reproduce it, and then fix it.

### Current OpenHands version

```bash
branch eval/24-sep
```

### Installation and Configuration

```bash
make build
```

### Model and Agent

_No response_

### Operating System

_No response_

### Reproduction Steps

_No response_

### Logs, Errors, Screenshots, and Additional Context

_No response_

### PR DESCRIPTION
**Pull request by xingyaoww**
- [X] Include this change in the Release Notes. If checked, you must provide an **end-user friendly** description for your change below

**End-user friendly description of the problem this fixes or functionality that this introduces**

Improve runtime error-handling on timeout

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

- Add few seconds to properly receive timeout error from client mostly fixed #4158

But the more complex story of #4158 originate from commands that cannot be killed with ONE ctrl+C. This PR:

- Add test that simulate the behavior of programs that cannot be killed with ctrl+C
- Improve runtime client to retry multiple times in sending ctrl+C - if eventually failed, use ctrl+Z ([SIGTSTP](https://superuser.com/questions/476873/what-is-effect-of-ctrl-z-on-a-unix-linux-application)) to stop the process, then `kill -9` to SIGTERM it.

---
**Link of any specific issues this addresses**

Fix https://github.com/All-Hands-AI/OpenHands/issues/4158

### CODE REVIEW

**Comment by xingyaoww** [comment]
Confirmed this has fixed the issue in #4158:
image

File: openhands/runtime/client/runtime.py [hunk 1]
```diff
@@ -418,7 +418,8 @@ def run_action(self, action: Action) -> Observation:
                 response = self.session.post(
                     f'{self.api_url}/execute_action',
                     json={'action': event_to_dict(action)},
-                    timeout=action.timeout,
+                    # wait a few more seconds to get the timeout error from client side
+                    timeout=action.timeout + 5,
```
Comment (tobitege) [inline openhands/runtime/client/runtime.py:422]: - Is the issue runtime-dependent or happening universally across all runtimes?
- Adding "+ 5" is a bit weird: shouldn't the default timeout have a minimum value then, that is at least 5 seconds (or what it is right now +5)? Putting the change at the source is easier to maintain than adding this all over the place? 🤔

Comment (xingyaoww) [reply openhands/runtime/client/runtime.py]: ah actually we can do it in the send request

**Comment by xingyaoww** [comment]
Actually the issue still wasn't fix.. (when EOF happens)

image

**Comment by xingyaoww** [comment]
I am able to reproduce this issue in a new test -- the issue arise when you have a process running that cannot be killed with ONE SIGINT, which will cause subsequent get return code `echo $?` to fail (bc you are not yet out of that program).

Will work on a fix tmr

**Comment by tobitege** [comment]
@xingyaoww why did running as openhands cause permission error in the runtime earlier? 
Shouldn't that have worked just as fine as for root? 🤔

**Comment by xingyaoww** [comment]
"chmod: changing permissions of 'resistant_script.sh': Operation not permitted\r\n\r\n[Py...enhands/workspace/rt_519509 $ ", command_id=-1, command='chmod +x resistant_script.sh', exit_code=1,

**Comment by tobitege** [comment]
@xingyaoww this is what Sonnet says about it, does that help?

-----

The problem lies in where you're trying to execute `chmod`. You're copying `resistant_script.sh` into the runtime's sandboxed environment (`runtime.config.workspace_mount_path_in_sandbox`) and then attempting to change its permissions *inside* the sandbox.  The sandbox likely has restricted permissions, preventing the `chmod` command from succeeding.

You should make the script executable *before* copying it into the sandbox.

```python:tests/runtime/test_bash.py
def test_process_resistant_to_multiple_sigint(temp_dir, box_class, run_as_openhands):
    # ... existing code ...

    with open(f'{temp_dir}/resistant_script.sh', 'w') as f:
        f.write(script_content)

    # Make the script executable *before* copying
    os.chmod(f'{temp_dir}/resistant_script.sh', 0o755)  # Or 0o775 depending on your needs

    runtime.copy_to(
        os.path.join(temp_dir, 'resistant_script.sh'),
        runtime.config.workspace_mount_path_in_sandbox,
    )

    # ... rest of the test ...

```
This change ensures that the script has the correct permissions before it's placed in the sandboxed environment, avoiding the permission error.  Using `os.chmod` directly is generally preferred over running `chmod` as a separate command.  The `0o755` permission gives execute permission to the owner, group, and others (read and execute for owner, read and execute for group and others).  Adjust this (e.g., to `0o700` or `0o775`) if you need different permissions.

**Review by tobitege — APPROVED** [review]
LGTM, not for the faint of heart all this! :) Thanks!!

**Comment by tobitege** [comment]
Just as I OK'ed this, the tests ran into the permission error again... 😔

**Pull request merged**